lib/kube/proxy/transport.go

```diff
@@ -0,0 +1,489 @@
+	// If the cluster is remote, the key is the teleport cluster name.
+	// If the cluster is local, the key is the teleport cluster name and the kubernetes
+	// cluster name: <teleport-cluster-name>/<kubernetes-cluster-name>.
+	key := transportCacheKey(sess)
+
+	// Check if the transport is cached.
+	cachedI, ok := f.cachedTransport.Get(key)
+	if ok {
+		if cached, ok := cachedI.(*httpTransport); ok {
+			if sess.upgradeToHTTP2 {
+				return cached.h2Transport, nil
+			}
+			return cached.h1Transport, nil
+		}
+	}
+
+	var httpTransport *httpTransport
+	var err error
+	if sess.teleportCluster.isRemote {
+		// If the cluster is remote, create a new transport for the remote cluster.
+		httpTransport, err = f.newRemoteClusterTransport(sess.teleportCluster.name)
+	} else if sess.kubeAPICreds != nil {
+		// If agent is running in agent mode, get the transport from the configured cluster
+		// credentials.
+		return sess.kubeAPICreds.getTransport(sess.upgradeToHTTP2), nil
+	} else if f.cfg.ReverseTunnelSrv != nil {
+		// If agent is running in proxy mode, create a new transport for the local cluster.
+		httpTransport, err = f.newLocalClusterTransport(sess.kubeClusterName)
+	} else {
+		return nil, trace.BadParameter("no reverse tunnel server or credentials provided")
+	}
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Cache the transport.
+	f.cachedTransport.Set(key, httpTransport, 5*time.Hour)
+	// Return the transport depending on whether HTTP/2 is enabled.
+	// Distinction is made because the SPDY protocol is not supported by HTTP/2
+	// and we must use HTTP/1.1 for it.
+	if sess.upgradeToHTTP2 {
+		return httpTransport.h2Transport, nil
+	}
+	return httpTransport.h1Transport, nil
+}
+
+// transportCacheKey returns a key used to cache transports.
+// If the cluster is remote, the key is the teleport cluster name.
+// If the cluster is local, the key is the teleport cluster name and the kubernetes
+// cluster name.
+func transportCacheKey(sess *clusterSession) string {
+	if sess.teleportCluster.isRemote {
+		return sess.teleportCluster.name
+	}
+	return filepath.Join(sess.teleportCluster.name, sess.kubeClusterName)
+}
+
+// wrapTransport wraps the provided transport with the Kubernetes transport config
+// if it is not nil.
+func wrapTransport(rt http.RoundTripper, transportConfig *transport.Config) (http.RoundTripper, error) {
+	if transportConfig == nil {
+		return rt, nil
+	}
+
+	wrapped, err := transport.HTTPWrappersForConfig(transportConfig, rt)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return enforceCloseIdleConnections(wrapped, rt), nil
+}
+
+// newTransport creates a new [http.Transport] with the provided dialer and TLS
+// config.
+// The transport is configured to use a connection pool and to close idle
+// connections after a timeout.
+func newTransport(dial dialContextFunc, tlsConfig *tls.Config) *http.Transport {
+	return &http.Transport{
```

Comment:
I think we don't use it because the `default.Transport` configures ProxyFromEnv and it also uses custom timeouts so it would require replacing almost all fields